mirrord/agent/src/main.rs

```diff
@@ -236,46 +263,69 @@ struct ClientConnectionHandler {
     tcp_stealer_api: TcpStealerApi,
     tcp_outgoing_api: TcpOutgoingApi,
     udp_outgoing_api: UdpOutgoingApi,
-    dns_sender: Sender<DnsRequest>,
+    dns_api: DnsApi,
     env: HashMap<String, String>,
 }
 
 impl ClientConnectionHandler {
-    #[allow(clippy::too_many_arguments)]
     /// Initializes [`ClientConnectionHandler`].
     pub async fn new(
         id: ClientId,
-        stream: TcpStream,
+        mut stream: Framed<TcpStream, DaemonCodec>,
         pid: Option<u64>,
         ephemeral: bool,
-        sniffer_command_sender: Sender<SnifferCommand>,
-        stealer_command_sender: Sender<StealerCommand>,
-        dns_sender: Sender<DnsRequest>,
+        bg_tasks: BackgroundTasks,
         env: HashMap<String, String>,
     ) -> Result<Self> {
         let file_manager = match pid {
             Some(_) => FileManager::new(pid),
             None if ephemeral => FileManager::new(Some(1)),
             None => FileManager::new(None),
         };
-        let stream = actix_codec::Framed::new(stream, DaemonCodec::new());
 
-        let (tcp_sender, tcp_receiver) = mpsc::channel(CHANNEL_SIZE);
+        let tcp_sniffer_api = match TcpSnifferApi::new(
+            id,
+            bg_tasks.sniffer_sender,
+            bg_tasks.sniffer_status,
+            CHANNEL_SIZE,
+        )
+        .await
+        {
+            Ok(api) => Some(api),
+            Err(e) => {
```

Comment:
You can do it with the [`futures::map_err`](https://docs.rs/futures/latest/futures/future/trait.TryFutureExt.html#method.map_err).

Reply:
> changed the `map_err(...).ok()` chain to this because there's no way to have an async closure in `map_err`, right?
Yup.

> You can do it with the [`futures::map_err`](https://docs.rs/futures/latest/futures/future/trait.TryFutureExt.html#method.map_err).
Not really, this method takes a simple closure :(. I do some async stuff when handling the error

Reply:
I tried, this just returns a `Result<TcpSnifferApi, [some anonymous async]>`. It doesn't run the async closure

Reply:
Oh, nvm then.